app/gui/src/presenter/graph/call_stack.rs

```diff
@@ -19,21 +20,178 @@ use ide_view as view;
+            let _transaction = main_module
+                .undo_redo_repository()
+                .open_ignored_transaction_or_ignore_current("Updating call stack metadata");
+            let new_call_stack = controller.call_stack();
+            let result = main_module.update_project_metadata(|metadata| {
+                metadata.call_stack = new_call_stack;
+            });
+            if let Err(error) = result {
+                // We cannot really do anything when updating metadata fails.
+                // Can happen in improbable case of serialization failure.
+                error!("Failed to store an updated call stack: {error}");
+            }
+        }
+    }
+
+    // === Stack History API ===
+    // Methods that modify the stack history. Note that these methods will not propagate the
+    // changes to the controller. They should be used to update the stack history, which is needed
+    // to translate the breadcrumbs selection to the call stack.
+
+    /// We enter a new node. If we enter a node that was not previous part of the stack history, we
+    /// discard deeper levels, otherwise we keep the deeper levels and just update the selection.
+    fn update_stack_history_on_node_enter(
+        &self,
+        controller_stack_size: usize,
+        new_call: LocalCall,
+    ) {
+        let new_item_index = controller_stack_size + 1;
+        let new_call = view::project_view_top_bar::LocalCall {
+            call:       new_call.call,
+            definition: new_call.definition.into(),
+        };
+        if self.stack_history.borrow().get(new_item_index) != Some(&new_call) {
+            self.stack_history.borrow_mut().truncate(new_item_index);
+            self.stack_history.borrow_mut().push(new_call);
+        }
+    }
+
+    // === UI Control API ===
+    // Methods that modify the UI state of the breadcrumbs. Note that they will always propagate
+    // the changes to the controller. These should be exclusivity used to update the state of the
+    // call stack. They will result in the breadcrumbs view being updated, which in turn will
+    // result in the methods from the `UI Driven API` being called to update the controller.
+
+    /// Add a part of the call stack. Note that this will check if the stack is already in the
+    /// breadcrumbs and if so, it will only update the selection.
+    fn add_stack_levels(
+        &self,
+        stack: Vec<view::project_view_top_bar::LocalCall>,
+        stack_pointer: usize,
+    ) {
+        if stack.is_empty() {
+            return;
+        }
+        let stack = stack.into_iter().map(|call| call.into()).collect_vec();
+        let breadcrumb_index = stack_pointer + 1;
+        self.view.top_bar().breadcrumbs.set_entries_from((stack, breadcrumb_index));
+    }
+
+    /// Remove the given number of levels from the stack. Will move the selection in the breadcrumbs
+    /// to the left.
+    fn shift_breadcrumb_selection_left(&self, count: usize) {
```

Comment:
The comment says we remove items from the stack, but I don't see it in the implementation. Is the comment correct?

Reply:
Clarified the comment. Modifying the selected item, will in turn trigger a change in the current active stack item.